app/src/main/java/org/mozilla/fenix/components/metrics/AppAllSourceStartTelemetry.kt

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.metrics
+
+import android.content.Intent
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
+import mozilla.components.support.utils.SafeIntent
+
+/**
+ * Tracks how the application was opened through [Event.AppOpenedAllSourceStartup]
+ * we only track the event if app is launching first time or was in background state because
+ * we don't want random intents with [Intent.ACTION_VIEW] to be tracked as app opened
+ * For example, Start app through APP_ICON -> run adb intent command to open a link.
+ * We want to ignore these adb commands since app is already on foreground.
+ */
+class AppAllSourceStartTelemetry(
+    private val metrics: MetricController
+) : LifecycleObserver {
+
+    private var wasApplicationInBackground = true
```

Comment:
When I see files with logic like this, where it's easy to flip a flag and mess up the behavior, I like to write tests to ensure I got it all right. Can we write a test for it?